### PR TITLE
Add Some API DSL Benchmarks

### PR DESCRIPTION
**Pull request by kitlangton**
Closes (some of) #1530

### CODE REVIEW

**Comment by kitlangton** [comment]
Oops. Didn't mean to request review. Still WIP.

File: zio-http-benchmarks/src/main/scala/zhttp.benchmarks/ApiBenchmark.scala [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+      collectHttpApp(request).repeatN(REPEAT_N)
+    }
+
+  // # Large Path
+
+  // API DSL
+  import In._
+
+  val largePathHttpApp = API
+    .get(In.literal("first") / int / "second" / int / "third" / int / "fourth" / int / "fifth" / int / "sixth" / int)
```
Comment (jdegoes) [inline zio-http-benchmarks/src/main/scala/zhttp.benchmarks/ApiBenchmark.scala]: This is a "deep path", which is great to benchmark!

In addition to the deep path, we should have a "broad path". There should be 1000 endpoints that arrange nicely into a tree: 10 wide by 10 deep. 

`O(n)` matchers will fair poorly in this case, but the prefix tree should fair quite well, all other things being equal.

The benchmark should hit a random API out of those 1000 endpoints just to be fair.

Comment (kitlangton) [reply zio-http-benchmarks/src/main/scala/zhttp.benchmarks/ApiBenchmark.scala]: We didn't get to 1000 endpoints yet, but on a semi-realistic set of broad routes, we're already starting to pull ahead:

```scala
[info] ApiBenchmark.benchmarkBroadTapirAkka        thrpt    4   78.794 ±  61.142  ops/s
[info] ApiBenchmark.benchmarkBroadTapirHttp4s      thrpt    4   86.456 ±  50.664  ops/s
[info] ApiBenchmark.benchmarkBroadZioApi           thrpt    4  667.506 ± 575.568  ops/s <-----
[info] ApiBenchmark.benchmarkBroadZioCollect       thrpt    4  119.175 ±  30.864  ops/s

[info] ApiBenchmark.benchmarkDeepPathTapirAkka     thrpt    4   39.212 ±  52.063  ops/s
[info] ApiBenchmark.benchmarkDeepPathTapirHttp4s   thrpt    4   42.098 ±   5.846  ops/s
[info] ApiBenchmark.benchmarkDeepPathZioApi        thrpt    4  433.606 ± 708.764  ops/s
[info] ApiBenchmark.benchmarkDeepPathZioCollect    thrpt    4  855.360 ± 342.678  ops/s

[info] ApiBenchmark.benchmarkSmallDataTapirAkka    thrpt    4   65.132 ±   5.211  ops/s
[info] ApiBenchmark.benchmarkSmallDataTapirHttp4s  thrpt    4   69.190 ±   3.894  ops/s
[info] ApiBenchmark.benchmarkSmallDataZioApi       thrpt    4  280.864 ±   5.790  ops/s
[info] ApiBenchmark.benchmarkSmallDataZioCollect   thrpt    4  681.826 ± 241.145  ops/s
```

Comment (kitlangton) [reply zio-http-benchmarks/src/main/scala/zhttp.benchmarks/ApiBenchmark.scala]: The `Collect` is using zio-http's `Http.collectZIO[Request]` with pattern matching. I think we should be able to match/beat that in the other benchmarks after further optimization (and pull even further ahead in the most realistic "broad" benchmarks).

**Comment by kitlangton** [comment]
#1546 built atop this PR and has been merged, so I'm closing this one.

**Pull request closed without merging**